File: ios/google_navigation_flutter/Sources/google_navigation_flutter/GoogleMapsNavigationSessionMessageHandler.swift
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

class GoogleMapsNavigationSessionMessageHandler: NavigationSessionApi {
  init(navigationSessionEventApi: NavigationSessionEventApi,
       viewRegistry: GoogleMapsNavigationViewRegistry) {
    GoogleMapsNavigationSessionManager.shared.setSessionEventApi(
      navigationSessionEventApi: navigationSessionEventApi
    )
    GoogleMapsNavigationSessionManager.shared.setViewRegistry(
      viewRegistry: viewRegistry
    )
  }

  /// General SDK functionality
  func showTermsAndConditionsDialog(title: String, companyName: String,
                                    shouldOnlyShowDriverAwarenessDisclaimer: Bool,
                                    completion: @escaping (Result<Bool, Error>) -> Void) {
    if shouldOnlyShowDriverAwarenessDisclaimer {
      // TODO: Disable driver awareness disclaimer on iOS due to the bug in the native side SDK
      completion(Result.failure(GoogleMapsNavigationSessionManagerError.notSupported))
      return
    }

    GoogleMapsNavigationSessionManager.shared.showTermsAndConditionsDialog(
      title: title,
      companyName: companyName,
      shouldOnlyShowDriverAwarenessDisclaimer: shouldOnlyShowDriverAwarenessDisclaimer
    ) { termsAccepted in
      completion(Result.success(termsAccepted))
    }
  }

  func areTermsAccepted() throws -> Bool {
    GoogleMapsNavigationSessionManager.shared.areTermsAccepted()
  }

  func resetTermsAccepted() throws {
    try GoogleMapsNavigationSessionManager.shared.resetTermsAccepted()
  }

  func getNavSDKVersion() -> String {
    GoogleMapsNavigationSessionManager.shared.getNavSDKVersion()
  }

  func createNavigationSession(abnormalTerminationReportingEnabled: Bool,
                               // taskRemovedBehaviourValue is Android only value and not used on
                               // iOS.
                               behavior: TaskRemovedBehaviorDto,
                               completion: @escaping (Result<Void, Error>) -> Void) {
    do {
      try GoogleMapsNavigationSessionManager.shared
        .createNavigationSession(abnormalTerminationReportingEnabled)
      completion(.success(()))
    } catch {
      completion(.failure(error))
    }
  }

  func isInitialized() -> Bool {
    GoogleMapsNavigationSessionManager.shared.isInitialized()
  }

  func cleanup() throws {
    try GoogleMapsNavigationSessionManager.shared.cleanup()
  }

  /// Navigation actions
  func isGuidanceRunning() throws -> Bool {
    try GoogleMapsNavigationSessionManager.shared.isGuidanceRunning()
  }

  func startGuidance() throws {
    try GoogleMapsNavigationSessionManager.shared.startGuidance()
  }

  func stopGuidance() throws {
    try GoogleMapsNavigationSessionManager.shared.stopGuidance()
  }

  func setDestinations(destinations: DestinationsDto,
                       completion: @escaping (Result<RouteStatusDto, Error>) -> Void) {
    GoogleMapsNavigationSessionManager.shared.setDestinations(
      destinations: destinations,
      completion: completion
    )
  }

  func clearDestinations() throws {
    try GoogleMapsNavigationSessionManager.shared.clearDestinations()
  }

  func continueToNextDestination() throws -> NavigationWaypointDto? {
    try GoogleMapsNavigationSessionManager.shared.continueToNextDestination()
  }

  func getCurrentTimeAndDistance() throws -> NavigationTimeAndDistanceDto {
    try GoogleMapsNavigationSessionManager.shared.getCurrentTimeAndDistance()
  }

  func setAudioGuidance(settings: NavigationAudioGuidanceSettingsDto) throws {
    try GoogleMapsNavigationSessionManager.shared.setAudioGuidance(settings: settings)
  }

  /// Simulation
  func simulateLocationsAlongExistingRoute() throws {
    try GoogleMapsNavigationSessionManager.shared.simulateLocationsAlongExistingRoute()
  }

  func setUserLocation(location: LatLngDto) throws {
    try GoogleMapsNavigationSessionManager.shared.setUserLocation(location: location)
  }

  func removeUserLocation() throws {
    try GoogleMapsNavigationSessionManager.shared.removeUserLocation()
  }

  func simulateLocationsAlongExistingRouteWithOptions(options: SimulationOptionsDto) throws {
    try GoogleMapsNavigationSessionManager.shared
      .simulateLocationsAlongExistingRouteWithOptions(options: options)
  }

  func simulateLocationsAlongNewRoute(waypoints: [NavigationWaypointDto],
                                      completion: @escaping (Result<
                                        RouteStatusDto,
                                        Error
                                      >)
                                        -> Void) {
    do {
      try GoogleMapsNavigationSessionManager.shared.simulateLocationsAlongNewRoute(
        waypoints: waypoints,
        completion: completion
      )
    } catch {
      completion(.failure(error))
    }
  }

  func simulateLocationsAlongNewRouteWithRoutingOptions(waypoints: [NavigationWaypointDto],
                                                        routingOptions: RoutingOptionsDto,
                                                        completion: @escaping (Result<
                                                          RouteStatusDto,
                                                          Error
                                                        >) -> Void) {
    do {
      try GoogleMapsNavigationSessionManager.shared
        .simulateLocationsAlongNewRouteWithRoutingOptions(
          waypoints: waypoints,
          routingOptions: routingOptions,
          completion: completion
        )
    } catch {
      completion(.failure(error))
    }
  }

  func simulateLocationsAlongNewRouteWithRoutingAndSimulationOptions(waypoints: [
    NavigationWaypointDto
  ],
  routingOptions: RoutingOptionsDto,
  simulationOptions: SimulationOptionsDto,
  completion: @escaping (Result<
    RouteStatusDto,
    Error
  >) -> Void) {
    do {
      try GoogleMapsNavigationSessionManager.shared
        .simulateLocationsAlongNewRouteWithRoutingAndSimulationOptions(
          waypoints: waypoints,
          routingOptions: routingOptions,
          simulationOptions: simulationOptions,
          completion: completion
        )
    } catch {
      completion(.failure(error))
    }
  }

  func pauseSimulation() throws {
    try GoogleMapsNavigationSessionManager.shared.pauseSimulation()
  }

  func resumeSimulation() throws {
    try GoogleMapsNavigationSessionManager.shared.resumeSimulation()
  }

  func allowBackgroundLocationUpdates(allow: Bool) throws {
    GoogleMapsNavigationSessionManager.shared.allowBackgroundLocationUpdates(allow: allow)
  }

  func setSpeedAlertOptions(options: SpeedAlertOptionsDto) throws {
    try GoogleMapsNavigationSessionManager.shared.setSpeedAlertOptions(options: options)
  }

  func getRouteSegments() throws -> [RouteSegmentDto] {
    try GoogleMapsNavigationSessionManager.shared.getRouteSegments()
  }

  func getTraveledRoute() throws -> [LatLngDto] {
    try GoogleMapsNavigationSessionManager.shared.getTraveledRoute()
  }

  func getCurrentRouteSegment() throws -> RouteSegmentDto? {
    try GoogleMapsNavigationSessionManager.shared.getCurrentRouteSegment()
  }

  /// Listeners
  func enableRoadSnappedLocationUpdates() throws {
    GoogleMapsNavigationSessionManager.shared.enableRoadSnappedLocationUpdates()
  }

  func disableRoadSnappedLocationUpdates() throws {
    GoogleMapsNavigationSessionManager.shared.disableRoadSnappedLocationUpdates()
  }

  func enableTurnByTurnNavigationEvents(numNextStepsToPreview: Int64?) throws {
    GoogleMapsNavigationSessionManager.shared
      .enableTurnByTurnNavigationEvents(numNextStepsToPreview: numNextStepsToPreview)
  }

  func disableTurnByTurnNavigationEvents() throws {
    GoogleMapsNavigationSessionManager.shared.disableTurnByTurnNavigationEvents()
  }

  func registerRemainingTimeOrDistanceChangedListener(remainingTimeThresholdSeconds: Int64,
                                                      remainingDistanceThresholdMeters: Int64) throws {
    GoogleMapsNavigationSessionManager.shared.registerRemainingTimeOrDistanceChangedListener(
      remainingTimeThresholdSeconds: remainingTimeThresholdSeconds,
      remainingDistanceThresholdMeters: remainingDistanceThresholdMeters
    )
  }
}

File: android/src/main/kotlin/com/google/maps/flutter/navigation/GoogleMapsNavigationSessionMessageHandler.kt
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.maps.flutter.navigation

import com.google.android.gms.maps.model.LatLng
import com.google.android.libraries.navigation.RoutingOptions
import com.google.android.libraries.navigation.SimulationOptions

class GoogleMapsNavigationSessionMessageHandler(
  private val sessionManager: GoogleMapsNavigationSessionManager
) : NavigationSessionApi {

  override fun createNavigationSession(
    abnormalTerminationReportingEnabled: Boolean,
    behavior: TaskRemovedBehaviorDto,
    callback: (Result<Unit>) -> Unit,
  ) {
    sessionManager.createNavigationSession(abnormalTerminationReportingEnabled, behavior, callback)
  }

  override fun isInitialized(): Boolean {
    return GoogleMapsNavigatorHolder.getInitializationState() ==
      GoogleNavigatorInitializationState.INITIALIZED
  }

  override fun cleanup(resetSession: Boolean) {
    sessionManager.cleanup(resetSession)
  }

  override fun showTermsAndConditionsDialog(
    title: String,
    companyName: String,
    shouldOnlyShowDriverAwarenessDisclaimer: Boolean,
    callback: (Result<Boolean>) -> Unit,
  ) {
    sessionManager.showTermsAndConditionsDialog(
      title,
      companyName,
      shouldOnlyShowDriverAwarenessDisclaimer,
      callback,
    )
  }

  override fun areTermsAccepted(): Boolean {
    return sessionManager.areTermsAccepted()
  }

  override fun resetTermsAccepted() {
    sessionManager.resetTermsAccepted()
  }

  override fun getNavSDKVersion(): String {
    return sessionManager.getNavSDKVersion()
  }

  override fun isGuidanceRunning(): Boolean {
    return sessionManager.isGuidanceRunning()
  }

  override fun startGuidance() {
    sessionManager.startGuidance()
  }

  override fun stopGuidance() {
    sessionManager.stopGuidance()
  }

  override fun setDestinations(
    destinations: DestinationsDto,
    callback: (Result<RouteStatusDto>) -> Unit,
  ) {
    val waypoints =
      destinations.waypoints.filterNotNull().map { Convert.convertWaypointFromDto(it) }
    val displayOptions = Convert.convertDisplayOptionsFromDto(destinations.displayOptions)
    val routingOptions =
      if (destinations.routingOptions != null) {
        Convert.convertRoutingOptionsFromDto(destinations.routingOptions)
      } else {
        RoutingOptions()
      }
    sessionManager.setDestinations(
      waypoints,
      routingOptions,
      displayOptions,
      destinations.routeTokenOptions,
    ) {
      if (it.isSuccess) {
        callback(Result.success(Convert.convertRouteStatusToDto(it.getOrThrow())))
      } else {
        val throwable = it.exceptionOrNull()
        if (throwable != null) {
          callback(Result.failure(throwable))
        }
      }
    }
  }

  override fun clearDestinations() {
    sessionManager.clearDestinations()
  }

  override fun continueToNextDestination(): NavigationWaypointDto? {
    val waypoint = sessionManager.continueToNextDestination()
    return if (waypoint != null) {
      Convert.convertWaypointToDto(waypoint)
    } else {
      null
    }
  }

  override fun getCurrentTimeAndDistance(): NavigationTimeAndDistanceDto {
    val timeAndDistance = sessionManager.getCurrentTimeAndDistance()
    return Convert.convertTimeAndDistanceToDto(timeAndDistance)
  }

  override fun setAudioGuidance(settings: NavigationAudioGuidanceSettingsDto) {
    val audioGuidanceSettings = Convert.convertAudioGuidanceSettingsToDto(settings)
    sessionManager.setAudioGuidance(audioGuidanceSettings)
  }

  override fun setSpeedAlertOptions(options: SpeedAlertOptionsDto) {
    val newOptions = Convert.convertSpeedAlertOptionsFromDto(options)
    sessionManager.setSpeedAlertOptions(newOptions)
  }

  override fun getRouteSegments(): List<RouteSegmentDto> {
    val routeSegments = sessionManager.getRouteSegments()
    return routeSegments.map { Convert.convertRouteSegmentToDto(it) }
  }

  override fun getTraveledRoute(): List<LatLngDto> {
    val traveledRoute = sessionManager.getTraveledRoute()
    return traveledRoute.map { LatLngDto(it.latitude, it.longitude) }
  }

  override fun getCurrentRouteSegment(): RouteSegmentDto? {
    val currentRouteSegment = sessionManager.getCurrentRouteSegment()
    if (currentRouteSegment != null) {
      return Convert.convertRouteSegmentToDto(currentRouteSegment)
    }
    return null
  }

  override fun setUserLocation(location: LatLngDto) {
    sessionManager.setUserLocation(LatLng(location.latitude, location.longitude))
  }

  override fun removeUserLocation() {
    sessionManager.removeUserLocation()
  }

  override fun simulateLocationsAlongExistingRoute() {
    sessionManager.simulateLocationsAlongExistingRoute()
  }

  override fun simulateLocationsAlongExistingRouteWithOptions(options: SimulationOptionsDto) {
    sessionManager.simulateLocationsAlongExistingRouteWithOptions(
      SimulationOptions().speedMultiplier(options.speedMultiplier.toFloat())
    )
  }

  override fun simulateLocationsAlongNewRoute(
    waypoints: List<NavigationWaypointDto>,
    callback: (Result<RouteStatusDto>) -> Unit,
  ) {
    val convertedWaypoints = waypoints.map { Convert.convertWaypointFromDto(it) }
    sessionManager.simulateLocationsAlongNewRoute(convertedWaypoints) {
      if (it.isSuccess) {
        callback(Result.success(Convert.convertRouteStatusToDto(it.getOrThrow())))
      } else {
        val throwable = it.exceptionOrNull()
        if (throwable != null) {
          callback(Result.failure(throwable))
        }
      }
    }
  }

  override fun simulateLocationsAlongNewRouteWithRoutingOptions(
    waypoints: List<NavigationWaypointDto>,
    routingOptions: RoutingOptionsDto,
    callback: (Result<RouteStatusDto>) -> Unit,
  ) {
    val convertedWaypoints = waypoints.map { Convert.convertWaypointFromDto(it) }
    sessionManager.simulateLocationsAlongNewRouteWithRoutingOptions(
      convertedWaypoints,
      Convert.convertRoutingOptionsFromDto(routingOptions),
    ) {
      if (it.isSuccess) {
        callback(Result.success(Convert.convertRouteStatusToDto(it.getOrThrow())))
      } else {
        val throwable = it.exceptionOrNull()
        if (throwable != null) {
          callback(Result.failure(throwable))
        }
      }
    }
  }

  override fun simulateLocationsAlongNewRouteWithRoutingAndSimulationOptions(
    waypoints: List<NavigationWaypointDto>,
    routingOptions: RoutingOptionsDto,
    simulationOptions: SimulationOptionsDto,
    callback: (Result<RouteStatusDto>) -> Unit,
  ) {
    val convertedWaypoints = waypoints.map { Convert.convertWaypointFromDto(it) }
    sessionManager.simulateLocationsAlongNewRouteWithRoutingAndSimulationOptions(
      convertedWaypoints,
      Convert.convertRoutingOptionsFromDto(routingOptions),
      SimulationOptions().speedMultiplier(simulationOptions.speedMultiplier.toFloat()),
    ) {
      if (it.isSuccess) {
        callback(Result.success(Convert.convertRouteStatusToDto(it.getOrThrow())))
      } else {
        val throwable = it.exceptionOrNull()
        if (throwable != null) {
          callback(Result.failure(throwable))
        }
      }
    }
  }

  override fun pauseSimulation() {
    sessionManager.pauseSimulation()
  }

  override fun resumeSimulation() {
    sessionManager.resumeSimulation()
  }

  override fun allowBackgroundLocationUpdates(allow: Boolean) {
    throw RuntimeException("Should only be called by iOS application")
  }

  override fun enableRoadSnappedLocationUpdates() {
    sessionManager.enableRoadSnappedLocationUpdates()
  }

  override fun disableRoadSnappedLocationUpdates() {
    sessionManager.disableRoadSnappedLocationUpdates()
  }

  override fun enableTurnByTurnNavigationEvents(numNextStepsToPreview: Long?) {
    sessionManager.enableTurnByTurnNavigationEvents(numNextStepsToPreview?.toInt() ?: Int.MAX_VALUE)
  }

  override fun disableTurnByTurnNavigationEvents() {
    sessionManager.disableTurnByTurnNavigationEvents()
  }

  override fun registerRemainingTimeOrDistanceChangedListener(
    remainingTimeThresholdSeconds: Long,
    remainingDistanceThresholdMeters: Long,
  ) {
    sessionManager.registerRemainingTimeOrDistanceChangedListener(
      remainingTimeThresholdSeconds,
      remainingDistanceThresholdMeters,
    )
  }
}
